### PR TITLE
Vectorized channel threshold search

### PR DESCRIPTION
**Pull request by eladc-git**
## Pull Request Description:
This PR removes the for loop for each threshold search per channel and uses numpy vectorization capability.
For KL divergence error function, we keep the loop because this KL involves histogram functions which are not trivial to vectorize.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py [hunk 1]
```diff
@@ -143,6 +144,36 @@ def _kl_error_function(x: np.ndarray,
                                range_max=range_max)
 
 
+def _kl_error_function_wrapper(x: np.ndarray,
+                               range_min: np.ndarray,
+                               range_max: np.ndarray,
+                               n_bins: int = 2048,
+                               n_bits: int = 8) -> np.ndarray:
+    """
+    Compute the error function between a tensor to its quantized version per channel.
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py:153]: Computes the error function between a tensor and its quantized version for each channel.
    The error is based on the KL-divergence between the distributions.
    The function uses a specified number of bins to compute the histogram of the float tensor.
    It requires the threshold and number of bits used for quantization to determine the histogram's boundaries and the number of quantized bins.

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py [hunk 2]
```diff
@@ -143,6 +144,36 @@ def _kl_error_function(x: np.ndarray,
...
+    histograms boundaries and the number of quantized bins.
+
+    Args:
+        x: Float tensor.
+        range_min: array of min bound on the quantization range.
+        range_max: array of max bound on the quantization range.
+        n_bins: Number of bins for the float histogram.
+        n_bits: Number of bits the quantized tensor was quantized by.
+
+    Returns:
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py:166]: An array containing the KL-divergence between the float and quantized histograms of the tensor for each channel.

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py [hunk 3]
```diff
@@ -143,6 +144,36 @@ def _kl_error_function(x: np.ndarray,
...
+    """
+    Compute the error function between a tensor to its quantized version per channel.
+    The error is computed based on the KL-divergence the distributions have.
+    Number of bins to use when computing the histogram of the float tensor is passed.
+    The threshold and number of bits that were used to quantize the tensor are needed to compute the
+    histograms boundaries and the number of quantized bins.
+
+    Args:
+        x: Float tensor.
+        range_min: array of min bound on the quantization range.
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py:161]: range_min: Array specifying the minimum bound of the quantization range for each channel.
        range_max: Array specifying the maximum bound of the quantization range for each channel.
        n_bins: Number of bins for the float histogram.
        n_bits: Number of bits used for quantization.

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py [hunk 4]
```diff
@@ -143,6 +144,36 @@ def _kl_error_function(x: np.ndarray,
...
+        n_bins: Number of bins for the float histogram.
+        n_bits: Number of bits the quantized tensor was quantized by.
+
+    Returns:
+        The KL-divergence of the float histogram and the quantized histogram of the tensors, per channel
+
+    """
+
+    error_list = []
+    for j in range(x.shape[0]):  # iterate all channels of the tensor.
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py:172]: is this always the axis of the channels? for all frameworks?

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py [hunk 5]
```diff
@@ -349,6 +381,7 @@ def get_threshold_selection_tensor_error_function(quantization_method: Quantizat
         quantization_method: Quantization method for threshold selection
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py:381]: quantization_method: Method used for selecting the quantization threshold.
quant_error_method: Type of error function requested.
p: P-norm to use for calculating the Lp-norm distance.
axis: Axis along which the operation has been performed.
norm: Indicates whether to normalize the result of the error function.
n_bits: Number of bits used to quantize the tensor.
signed: Indicates whether the input is signed.

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/power_of_two_selection.py [hunk 6]
```diff
@@ -55,8 +55,9 @@ def power_of_two_selection_tensor(tensor_data: np.ndarray,
         threshold = max_power_of_two(tensor_max, min_threshold)
     else:
         signed = True  # weights are always signed
+        axis = -1 if per_channel else None
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/power_of_two_selection.py:58]: I don't think this is the case for Pytorch. I think you need to use the fw_info

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/qparams_search.py [hunk 7]
```diff
@@ -726,9 +726,7 @@ def _error_function_wrapper(error_function: Callable,
         q_tensor: Numpy array with quantized tensor's content.
         in_params: Quantization params the tensor is quantized by (used in specific error functions only).
 
-    Returns: A list of error values per-channel for the quantized tensor, according to the error function.
+    Returns: A array of error values per-channel for the quantized tensor, according to the error function.
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/qparams_search.py:729]: Returns: An array of error values for each channel of the quantized tensor, as determined by the specified error function.

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/symmetric_selection.py [hunk 8]
```diff
@@ -58,7 +58,8 @@ def symmetric_selection_tensor(tensor_data: np.ndarray,
         threshold = get_init_threshold(min_threshold, tensor_max, per_channel)
     else:
         signed = True  # weights are always signed
-        error_function = get_threshold_selection_tensor_error_function(QuantizationMethod.SYMMETRIC, quant_error_method, p, norm=False, n_bits=n_bits, signed=signed)
+        axis = -1 if per_channel else None
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/symmetric_selection.py:61]: I don't think this is the case for Pytorch. I think you need to use the fw_info

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/uniform_selection.py [hunk 9]
```diff
@@ -56,7 +56,8 @@ def uniform_selection_tensor(tensor_data: np.ndarray,
     if quant_error_method == qc.QuantizationErrorMethod.NOCLIPPING:
         mm = tensor_min, tensor_max
     else:
-        error_function = get_threshold_selection_tensor_error_function(QuantizationMethod.UNIFORM, quant_error_method, p, norm=False)
+        axis = -1 if per_channel else None
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/common/quantization/quantization_params_generation/uniform_selection.py:59]: I don't think this is the case for Pytorch. I think you need to use the fw_info

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py (hunk 4 above)
Comment (eladc-git) [reply model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py:172]: yes. We make it that way in the quantization search.

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/power_of_two_selection.py (hunk 6 above)
Comment (eladc-git) [reply model_compression_toolkit/core/common/quantization/quantization_params_generation/power_of_two_selection.py:58]: Numpy. This is in the common section.

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/symmetric_selection.py (hunk 8 above)
Comment (eladc-git) [reply model_compression_toolkit/core/common/quantization/quantization_params_generation/symmetric_selection.py:61]: numpy

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/uniform_selection.py (hunk 9 above)
Comment (eladc-git) [reply model_compression_toolkit/core/common/quantization/quantization_params_generation/uniform_selection.py:59]: numpy

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py (hunk 1 above)
Comment (eladc-git) [reply model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py:153]: I didn't change the descriptions in this PR, but I can change them.

**Review by lior-dikstein — APPROVED** [review]

**Pull request merged**